File: project/.sdk/tm/lua/utility/make_url.lua
-- ProjectName SDK utility: make_url

local vs = require("utility.struct.struct")

local function make_url_util(ctx)
  local spec = ctx.spec
  local result = ctx.result

  if spec == nil then
    return "", ctx:make_error("url_no_spec",
      "Expected context spec property to be defined.")
  end
  if result == nil then
    return "", ctx:make_error("url_no_result",
      "Expected context result property to be defined.")
  end

  local url = vs.join({ spec.base, spec.prefix, spec.path, spec.suffix }, "/", true)
  local resmatch = {}

  local param_items = vs.items(spec.params)
  if param_items ~= nil then
    for _, item in ipairs(param_items) do
      local key = item[1]
      local val = item[2]
      if val ~= nil and type(key) == "string" then
        local placeholder = "{" .. key .. "}"
        local val_str = type(val) == "string" and val or tostring(val)
        local encoded = vs.escurl(val_str)
        -- Plain string replacement (not pattern-based)
        local i, j = url:find(placeholder, 1, true)
        while i do
          url = url:sub(1, i - 1) .. encoded .. url:sub(j + 1)
          i, j = url:find(placeholder, i + #encoded, true)
        end
        resmatch[key] = val
      end
    end
  end

  result.resmatch = resmatch

  return url, nil
end

return make_url_util

File: project/.sdk/tm/py/utility/make_url.py
# ProjectName SDK utility: make_url

from __future__ import annotations
from utility.voxgig_struct import voxgig_struct as vs


def make_url_util(ctx):
    spec = ctx.spec
    result = ctx.result

    if spec is None:
        return "", ctx.make_error("url_no_spec",
            "Expected context spec property to be defined.")
    if result is None:
        return "", ctx.make_error("url_no_result",
            "Expected context result property to be defined.")

    url = vs.join([spec.base, spec.prefix, spec.path, spec.suffix], "/", True)
    resmatch = {}

    param_items = vs.items(spec.params)
    if param_items is not None:
        for item in param_items:
            key = item[0]
            val = item[1]
            if val is not None and isinstance(key, str):
                val_str = val if isinstance(val, str) else str(val)
                encoded = vs.escurl(val_str)
                url = url.replace("{" + key + "}", encoded)
                resmatch[key] = val

    result.resmatch = resmatch

    return url, None
